code_search/app/code_search/nmslib/search_engine.py

```diff
@@ -0,0 +1,69 @@
+import json
+import requests
+import nmslib
+import numpy as np
+from tensor2tensor import problems  # pylint: disable=unused-import
+from code_search.t2t.query import get_encoder_decoder, encode_query
+
+
+class CodeSearchEngine:
+  """This is a utility class which takes an nmslib
+  index file and a data file to return data from"""
+  def __init__(self, problem: str, data_dir: str, serving_url: str,
+               index_file: str):
+    self._serving_url = serving_url
+    self._problem = problem
+    self._data_dir = data_dir
+    self._index_file = index_file
+
+    self.index = CodeSearchEngine.nmslib_init()
+    self.index.loadIndex(index_file)
+
+  def embed(self, query_str):
+    """This function gets the vector embedding from
+    the target inference server. The steps involved are
+    encoding the input query and decoding the responses
+    from the TF Serving service
+    TODO(sanyamkapoor): This code is still under construction
+    and only representative of the steps needed to build the
+    embedding
+    """
+    encoder, decoder = get_encoder_decoder(self._problem, self._data_dir)
+    encoded_query = encode_query(encoder, query_str)
+    data = {"instances": [{"input": {"b64": encoded_query}}]}
+
+    response = requests.post(url=self._serving_url,
+                             headers={'content-type': 'application/json'},
+                             data=json.dumps(data))
+
+    result = response.json()
+    for prediction in result['predictions']:
+      prediction['outputs'] = decoder.decode(prediction['outputs'])
```

Comment:
Why are you decoding it? Isn't the point of the TF serving server to embed the the search query in the space? So doesn't the TF serving model return the embedded vector?

Reply:
I wrote this part w.r.t another model and is representative of what "might" need to be done. From today, I am going to start investigating the graph (a simple single layer graph which takes input and gives some embedding out).